samcli/lib/pipeline/bootstrap/resource.py

```diff
@@ -0,0 +1,133 @@
+""" Represents AWS resource"""
+from typing import List, Optional
+
+
+class ARNParts:
+    """
+    Decompose a given ARN into its parts https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html
+
+    Attributes
+    ----------
+    partition: str
+        the partition part(AWS, aws-cn or aws-us-gov)  of the ARN
+    service: str
+        the service part(S3, IAM, ECR, ...etc) of the ARN
+    region: str
+        the AWS region part(us-east-1, eu-west-1, ...etc) of the ARN
+    account-id: str
+        the account-id part of the ARN
+    resource-id: str
+        the resource-id part of the ARN
+    resource-type: str
+        the resource-type part of the ARN
+    """
+
+    def __init__(self, arn: str) -> None:
+        if not isinstance(arn, str):
+            raise ValueError(f"Invalid ARN ({arn}) is not a String")
+
+        parts = arn.split(":")
+        try:
+            self.partition: str = parts[1]
+            self.service: str = parts[2]
+            self.region: str = parts[3]
+            self.account_id: str = parts[4]
+            self.resource_id: str = parts[5]
```

Comment:
This looks simplier

```suggestion
            [self.partition, self.service, self.region......] = parts[1]
```

Reply:
simpler like better readability?

Reply:
this is python's way to assign elements in a list to different variables. Yeah, it helps with the readability and it is less prune to typos because there are no need to maintain the index

Reply:
IMO, the current assignment is more readable in this case especially with the type-hint inplace. Regarding index being error prone, it is the same issue if we swapped two variables.

Reply:
another downside of doing it using index: we cannot verify the length of `parts`. Take this as an example, if the format is unexpected and making `parts` have 7 elements, this code will still run without exception.

Reply:
```py
class ARNParts:

    partition: str
    service: str
    region: str
    account_id: str
    resource_id: str

    def __init__(self, arn: str) -> None:
        if not isinstance(arn, str):
            raise ValueError(f"Invalid ARN ({arn}) is not a String")

        parts = arn.split(":")
        try:
            [self.partition, self.service, self.region, self.account_id, self.resource_id] = parts
        except IndexError as ex:
            raise ValueError(f"Invalid ARN ({arn})") from ex
```

Here is a more concrete example why this notation is better: it can catch fires when `parts` has more elements than expected:

<img width="548" alt="image" src="https://user-images.githubusercontent.com/3804518/115311518-9ecc0f00-a124-11eb-8938-c2b7c5f99beb.png">